### PR TITLE
Add data names to convergence and iteration file

### PR DESCRIPTION
**Pull request by KyleDavisSA**
This PR aims to resolve https://github.com/precice/precice/issues/555. The previous output went from dataID 0 and upwards:

> resNorm(0)  resNorm(1)  resNorm(2)

printing `convMeasure.data->getID()` does not print the data ID in this order. The different data ID's are looped through by 

https://github.com/precice/precice/blob/fe00f53e059f57e6c390d4fcfa60a25da896fe1c/src/cplscheme/BaseCouplingScheme.cpp#L695-L703

An example output was generated for the cylinderFlap example

```
Timestep  Iterations  ResNorm(3)  ResNorm(2)  
1  1  1.0000000000000000  1.0000000000000000  
1  2  1.0223306624150461  1.2125016658484906  
1  3  1.0104960419007956  1.5949903615016618  
1  4  0.9907038391847834  6.9682477895700625
```
The same dataID is also found when using 
`sstm << "ResNorm(" << _convergenceMeasures[i].data->getID() << ");`

For now, it prints the dataID and not the dataName.

### CODE REVIEW

**Comment by uekerman** [comment]
The order of convergence measures is determined by their order in the config. And I guess that's the expected behavior. No need to sort them by ID. 
`_convergenceMeasures[i].data->getName()` should give you the name.

File: src/cplscheme/BaseCouplingScheme.cpp [hunk 1]
```diff
@@ -676,34 +676,35 @@ void BaseCouplingScheme::initializeTXTWriters()
...
         std::stringstream sstm, sstm2;
-        sstm << "avgConvRate(" << i << ")";
-        sstm2 << "resNorm(" << i << ")";
+        sstm << "AvgConvRate(" << convMeasure.data->getID() << ")";
+        sstm2 << "ResNorm(" << convMeasure.data->getID() << ")";
         _iterationsWriter->addData(sstm.str(), io::TXTTableWriter::DOUBLE);
         _convergenceWriter->addData(sstm2.str(), io::TXTTableWriter::DOUBLE);
       }
-      _iterationsWriter->addData("deleted_Columns", io::TXTTableWriter::INT);
+      _iterationsWriter->addData("deletedColumns", io::TXTTableWriter::INT);
```
Comment (uekerman) [inline src/cplscheme/BaseCouplingScheme.cpp]: Wouldn't `"DeletedColumns"` be more consistent?

**Comment by KyleDavisSA** [comment]
Previously, the output was 
`resNorm(0) resNorm(1) ...`

If you print the dataID for each data name in the solvers, you find for the Fluid solver:
```
Forces = 0
Displacements = 1
```
and for the Solid solver:
```
Forces = 2
Displacements = 3
```

As an example, the current convergence file will print the dataID by looping through variable `i`:
https://github.com/precice/precice/blob/c0548c720ca681b3e1bac75e8565f14b49b8a8ad/src/cplscheme/BaseCouplingScheme.cpp#L598
```
Timestep  Iterations  ResNorm(0)  ResNorm(1)  
1  1  1.0000000000000000  1.0000000000000000  
1  2  1.0227066777948459  1.2158700458329670  
1  3  1.0113675981800501  1.7151706495425945  
```
By printing the dataID for the associated resNorm measure will print:
https://github.com/precice/precice/blob/3f732ff84066e1cc89b3bd145d10613b3cf34b6d/src/cplscheme/BaseCouplingScheme.cpp#L598
```
Timestep  Iterations  ResNorm(3)  ResNorm(2)  
1  1  1.0000000000000000  1.0000000000000000  
1  2  1.0227066777948459  1.2158700458329670  
1  3  1.0113675981800501  1.7151706495425945 
```
And by printing the dataName:
```
Timestep  Iterations  ResNorm(Displacements)  ResNorm(Forces)  
1  1  1.0000000000000000  1.0000000000000000  
1  2  1.0227066777948459  1.2158700458329670  
1  3  1.0113675981800501  1.7151706495425945 
```
Does the current implementation printout the resNorm for dataID `3` and names it as `ResNorm(0)`?

**Comment by uekerman** [comment]
Not sure about which ID is used exactly in the convergence measure, but we should actually never output IDs. So, the current state is certainly sub-optimal.   
```
ResNorm(Displacements)  ResNorm(Forces)
```
does the job.

File: src/cplscheme/BaseCouplingScheme.cpp [hunk 2]
```diff
@@ -676,34 +676,34 @@ void BaseCouplingScheme::initializeTXTWriters()
...
       _convergenceWriter->addData("Timestep", io::TXTTableWriter::INT);
-      _convergenceWriter->addData("Iteration", io::TXTTableWriter::INT);
+      _convergenceWriter->addData("Iterations", io::TXTTableWriter::INT);
     }
 
-    int i = -1;
     if (not doesFirstStep()) {
-      for (ConvergenceMeasure &convMeasure : _convergenceMeasures) {
-        i++;
+      for (size_t i = 0; i < _convergenceMeasures.size(); i++) {
```
Comment (MakisH) [inline src/cplscheme/BaseCouplingScheme.cpp]: Why converting from the ranged-for (previous) to the "classic" for-loop? We don't need the counter `i` anywhere and a ranged-for is safer.

File: src/cplscheme/BaseCouplingScheme.cpp [hunk 3]
```diff
@@ -676,34 +676,34 @@ void BaseCouplingScheme::initializeTXTWriters()
...
 
-    _iterationsWriter->addData("Timesteps", io::TXTTableWriter::INT);
-    _iterationsWriter->addData("Total_Iterations", io::TXTTableWriter::INT);
+    _iterationsWriter->addData("Timestep", io::TXTTableWriter::INT);
+    _iterationsWriter->addData("TotalIterations", io::TXTTableWriter::INT);
     _iterationsWriter->addData("Iterations", io::TXTTableWriter::INT);
     if (hasCoarseModelOptimization) {
-      _iterationsWriter->addData("Total_Iterations_Surrogate_Model", io::TXTTableWriter::INT);
-      _iterationsWriter->addData("Iterations_Surrogate_Model", io::TXTTableWriter::INT);
+      _iterationsWriter->addData("TotalIterations_SurrogateModel", io::TXTTableWriter::INT);
```
Comment (MakisH) [inline src/cplscheme/BaseCouplingScheme.cpp]: I understand you are trying to nicely separate between `TotalIterations` and `TotalIterations_SurrogateModel`, but wouldn't just `TotalIterationsSurrogateModel` be good (and consistent) as well? Does [clang-format](http://xgm.de/precice/docs/develop/tooling.html) suggest any change?

File: src/cplscheme/BaseCouplingScheme.cpp [hunk 4]
```diff
@@ -718,33 +718,35 @@ void BaseCouplingScheme::advanceTXTWriters()
...
     if (hasCoarseModelOptimization) {
-      _iterationsWriter->writeData("Total_Iterations_Surrogate_Model", _totalIterationsCoarseOptimization);
-      _iterationsWriter->writeData("Iterations_Surrogate_Model", _iterationsCoarseOptimization);
+      _iterationsWriter->writeData("TotalIterations_SurrogateModel", _totalIterationsCoarseOptimization);
+      _iterationsWriter->writeData("Iterations_SurrogateModel", _iterationsCoarseOptimization);
     }
     int converged = _iterations < _maxIterations ? 1 : 0;
     _iterationsWriter->writeData("Convergence", converged);
 
+    int i = -1;
```
Comment (MakisH) [inline src/cplscheme/BaseCouplingScheme.cpp]: We don't need `i` anywhere, right?

File: src/cplscheme/BaseCouplingScheme.cpp [hunk 5]
```diff
@@ -575,7 +575,7 @@ bool BaseCouplingScheme::measureConvergence(
   PRECICE_ASSERT(_convergenceMeasures.size() > 0);
   if (not utils::MasterSlave::isSlave()) {
     _convergenceWriter->writeData("Timestep", _timesteps);
-    _convergenceWriter->writeData("Iteration", _iterations);
+    _convergenceWriter->writeData("Iterations", _iterations);
```
Comment (MakisH) [inline src/cplscheme/BaseCouplingScheme.cpp]: From the results, I understand that this should be `Iteration` (as before), not `Iterations`. For example:

```
Timestep  Iterations  ResNorm(Displacements0)  ResNorm(Forces0)  
1  1  0.0003191866323491  1.0000000000000000  
1  2  0.0003159904961397  0.0000149565768695  
1  3  0.0000000004065084  0.0014792261490834  
2  1  0.0007445657741119  0.2056384940762610  
2  2  0.0000002404722286  0.0039091692413695  
3  1  0.0005485913384467  0.1860456395794740  
3  2  0.0000000664794586  0.0033499436260277  
```
Looking at the `*-convergence.log` file I usually look if the residual norm decreases with every iteration. `Iterations` is not what is meant here, I guess (even if the variable is named `_iterations`).

Am I misunderstanding something?

Comment (KyleDavisSA) [reply src/cplscheme/BaseCouplingScheme.cpp]: I think that would be right. I vote for keep `TotalIterations` the same and make it `Iteration` (singular, refering to the result of a specific iteration). I don't think changing the variable `_iterations` is necessary.

File: src/cplscheme/BaseCouplingScheme.cpp (hunk 4 above)
Comment (KyleDavisSA) [reply src/cplscheme/BaseCouplingScheme.cpp]: `i` is used in calculation of the residual below (line742)

Comment (MakisH) [reply src/cplscheme/BaseCouplingScheme.cpp]: Then it's ok, I guess! :-)

**Comment by fsimonis** [comment]
@KyleDavisSA the cplscheme tests fail as their data names don't match on write.

**Pull request merged**